### ISSUE TITLE
aws_cdk.custom_resources: AWSCDKCfnUtilsProviderCustomResourceProviderRole is created and is not part of the CDK path

### ISSUE THREAD

**Issue by leoneperdigao** [aws/aws-cdk#22972]
### Describe the bug

While using CDK Custom resources, a ghost role is created under the hood which is not part of the CDK path. By not being part of the CDK path, it is impossible to apply permissions boundary via Aspects or by whatever means.

### Expected Behavior

The generated role is part of the path as all the other resources and, therefore, suitable to aspects application.

### Current Behavior

After the synthesis, the following role is generated with the path `"my-stack/AWSCDKCfnUtilsProviderCustomResourceProvider/Role"`. However it is not possible whatsoever to find such resource in the app tree.  The resource is only available after synthesis.

```json
"AWSCDKCfnUtilsProviderCustomResourceProviderRoleFE0EE867": {
   "Type": "AWS::IAM::Role",
   "Properties": {
    "AssumeRolePolicyDocument": {
     "Version": "2012-10-17",
     "Statement": [
      {
       "Action": "sts:AssumeRole",
       "Effect": "Allow",
       "Principal": {
        "Service": "lambda.amazonaws.com"
       }
      }
     ]
    },
    "ManagedPolicyArns": [
     {
      "Fn::Sub": "arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
     }
    ]
   },
   "Metadata": {
    "aws:cdk:path": "my-stack/AWSCDKCfnUtilsProviderCustomResourceProvider/Role"
   }
  },
```

### Reproduction Steps
Aspect:

```python

"""Module for CDK Aspects."""

import jsii
from aws_cdk import Aws, IAspect, CfnResource


@jsii.implements(IAspect)
class IamRolePermissionsBoundaryAspect:  # pylint: disable=too-few-public-methods
    """CDK Aspect to add the permissions boundary to every IAM Role."""

    def __init__(self):
        self.permissions_boundary_arn = (
            f"arn:aws:iam::{Aws.ACCOUNT_ID}:policy/lpe-base-permissions-boundary"
        )

    def visit(self, node):
        """Visit every node, check if it is a role, update if it is."""

        if CfnResource.is_cfn_resource(node) and node.cfn_resource_type == "AWS::IAM::Role":
            # The role does not have a permissions boundary, or it is not
            # the lpe-base-permissions-boundary, so let's add it.
            node.add_property_override("PermissionsBoundary", self.permissions_boundary_arn)

```

App:

```python
#!/usr/bin/env python
import aws_cdk as cdk
from aspects.iam_role_permissions_boundary import IamRolePermissionsBoundaryAspect

app = cdk.App()
    app_stack = MyStack(app, "management-app")
    Aspects.of(app).add(IamRolePermissionsBoundaryAspect())
```

### Additional Information/Context

Related issues:
#15572
#3242
#3310

### CDK CLI Version

2.50.0

### Node.js Version

16

### OS

macOS

### Language

Python

### Language Version

3.9

**Comment by peterwoodworth** [comment]
Can you provide reproduction steps for creating this role? You've only supplied the Aspect reproduction steps so I'm not certain how this role is being generated

**Comment by nocquidant** [comment]
Same issue found here: my organisation doesn't allow role creation without a boundary attached.
Or this ghost-role (AWSCDKCfnUtilsProviderCustomResourceProviderRoleFE0EE867) is not amendable...

Error: 

```
9:36:27 AM | CREATE_FAILED        | AWS::IAM::Role                 | AWSCDKCfnUtilsProv...oviderRoleFE0EE867
Resource handler returned message: "Encountered a permissions error performing a tagging operation, please add required tag permissions. See https://re
post.aws/knowledge-center/cloudformation-tagging-permission-error for how to resolve. Resource handler returned message: "User: arn:aws:sts::2997035557
57:assumed-role/cf-app-noc-deployer-role/AWSCloudFormation is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::299703555757:role/noc
-dmo-appsync-vpce-dev--AWSCDKCfnUtilsProviderCus-5WrO4F59gAcl because no identity-based policy allows the iam:CreateRole action (Service: Iam, Status C
ode: 403, Request ID: 51f6419b-9d7a-4611-b0be-912935de5f0d)"" (RequestToken: 348afcbd-35f2-ecef-c439-617cf13f7225, HandlerErrorCode: UnauthorizedTaggin
gOperation)
```

Stack:

```
export class VpceStack extends BaseStack {

    constructor(scope: Construct, id: string, props: BaseStackProps) {
        super(scope, id, props);

        const policies = new LambdaPolicies(this.locals);

        const defaultRole = new Role(this, this.id('lambda-basic-role'), {
            roleName: this.name('basic-role'),
            assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
            managedPolicies: policies.managedPolicies(this),
            permissionsBoundary: this.project.defaultLambdaBoundary(),
            inlinePolicies: policies.inlinePolicies(this.project.cmk().keyArn),
        });

        // Retrieve VPC Restricted
        const vpcRestricted = Vpc.fromLookup(this, this.id("vpc-restricted"), {
            isDefault: false,
            vpcId: "vpc-02de2f0d152316a2b"
        });

        // Security groups
        const vpceAppsyncSG = new ec2.SecurityGroup(
            this, this.id("vpce-appsync-sg"),
            {
                vpc: vpcRestricted,
                securityGroupName: this.name("vpce-appsync-sg"),
                description: "Allow Access to AppSync",
                allowAllOutbound: true,
            },
        );

        // Create VPC Endpoint for AppSync with SG previously created
        const vpceAppSync = vpcRestricted.addInterfaceEndpoint(this.id("vpc-endpoint-appsync"),
            {
                service: ec2.InterfaceVpcEndpointAwsService.APP_SYNC,
                privateDnsEnabled: true,
                subnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
                securityGroups: [vpceAppsyncSG],
            },
        );

        // find IpAddresses of VPCEndpoint attach to a subnet
        const getEndpointPrivateIpAddress = (index: number) => {
            const privateIpAddressField = `NetworkInterfaces.${index}.PrivateIpAddress`;
            const resource = new AwsCustomResource(this, `GetEndpointIp${index}`, {
                role: defaultRole,
                onUpdate: {
                    service: "EC2",
                    action: "describeNetworkInterfaces",
                    outputPaths: [privateIpAddressField],
                    parameters: {
                        NetworkInterfaceIds: vpceAppSync.vpcEndpointNetworkInterfaceIds,
                    },
                    physicalResourceId: PhysicalResourceId.of(privateIpAddressField),
                },
                policy: AwsCustomResourcePolicy.fromSdkCalls({
                    resources: AwsCustomResourcePolicy.ANY_RESOURCE,
                }),
            });
            return resource.getResponseField(privateIpAddressField);
        };

        this.output(this.id('EndpointPrivateIpAddress'), getEndpointPrivateIpAddress(0));
    }
}
```

**Comment by leoneperdigao** [comment]
> Can you provide reproduction steps for creating this role? You've only supplied the Aspect reproduction steps so I'm not certain how this role is being generated

That is the thing, this role is not created by me but by CDK itself when there is a need for Custom Resources. CDK will generate a Lambda function and this lambda requires an execution role. This role is the one that creates a problem, because the aspect is not applied timely. Like I mentioned, the role becomes "visible" only after synthesis, hence the Aspect is not applied properly.

**Comment by smiling-gopher** [comment]
@peterwoodworth 
The role, `AWSCDKCfnUtilsProviderCustomResourceProviderHandlerCF82AA57`, is created by CDK and not accessible via the node/construct tree.
Also, CDK Aspect has no effects on this role.
&nbsp;
This seems to affect the custom resource when using service `ec2` and action `describeNetworkInterfaces` then passing `vpcEndpointNetworkInterfaceIds` from a variable like so:

```
const vpc = new ec2.Vpc(this, 'MyVpc', {...});

const endpoint = vpc.addInterfaceEndpoint('MyEndpoint', {
        service: ec2.InterfaceVpcEndpointAwsService.EC2,
        privateDnsEnabled: true,
        subnets: {
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
        },
});

const customresource = new cr.AwsCustomResource(this, `NetworkInterfaces${i}PrivateIpAddress`, {
          onUpdate: {
            service: 'ec2',
            action: 'describeNetworkInterfaces',
            parameters: {
              NetworkInterfaceIds: endpoint.vpcEndpointNetworkInterfaceIds,
            },
            outputPaths: [privateAddressField],
            physicalResourceId: cr.PhysicalResourceId.of(privateAddressField),
          },
          policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
            resources: cr.AwsCustomResourcePolicy.ANY_RESOURCE,
          }),
});
```

**Comment by leoneperdigao** [comment]
@rix0rrr could you please reopen the issue? It still happens currently.

**Comment by imduchy** [comment]
We're facing exactly the same issue.

In our organization, all roles must have a permissions boundary and path assigned. We have Aspects taking care of patching auto-generated roles but the `Custom::AWSCDKOpenIdConnectProviderCustomResourceProvider/Role` and `AWSCDKCfnUtilsProviderCustomResourceProvider/Role` roles seem to be ignored.

For now, the only way to patch these roles is inside of the code

```py
oidc_cr: CustomResourceProvider = self.node.find_child("Custom::AWSCDKOpenIdConnectProviderCustomResourceProvider")
oidc_role: CfnResource = oidc_cr.node.find_child("Role")
oidc_role.add_override("Properties.Path", "/approles/cr/")
```

This is, however, not an ideal solution as we're patching in two different places instead of one.
